site/app/controllers/course/CourseMaterialsController.php

```diff
@@ -389,29 +389,74 @@ public function ajaxEditCourseMaterialsFiles(bool $flush = true): JsonResponse {
-                $dirs = explode("/", $path);
-                array_pop($dirs);
-                $path = implode("/", $dirs);
-                $path = FileUtils::joinPaths($path, urlencode("link-" . $_POST['link_title']));
+
+        if (isset($_POST['release_time']) && $_POST['release_time'] != '') {
+            $date_time = DateUtils::parseDateTime($_POST['release_time'], $this->core->getDateTimeNow()->getTimezone());
+            $course_material->setReleaseDate($date_time);
+        }
+
+        if (isset($_POST['file_path'])) {
+            $path = $course_material->getPath();
+            $upload_path = FileUtils::joinPaths($this->core->getConfig()->getCoursePath(), "uploads", "course_materials");
+            $requested_path = $_POST['file_path'];
+            $new_path = $upload_path . "/" . $requested_path;
+            if ($course_material->isFile() && $path !== $new_path) {
+                if (!$this->checkValidPath($requested_path, $upload_path)) {
+                    return JsonResponse::getErrorResponse("Invalid requested path");
+                }
+                $new_path = FileUtils::joinPaths($upload_path, $requested_path);
+                // check valid directory
+                $requested_path = explode("/", $requested_path);
+                if (count($requested_path) > 1) {
+                    $dir = explode("/", $new_path);
+                    array_pop($dir);
+                    $dir = implode("/", $dir);
+                    $cm = $this->core->getCourseEntityManager()->getRepository(CourseMaterial::class)
+                    ->findOneBy(['path' => $dir]);
+                    if ($cm == null) {
+                        return JsonResponse::getErrorResponse("The specified path does not exits. Please create folder(s) or move into existing folder");
+                    }
+                }
+                $tmp_course_material = $this->core->getCourseEntityManager()->getRepository(CourseMaterial::class)
+                        ->findOneBy(['path' => $new_path]);
+                if ($tmp_course_material !== null) {
+                    return JsonResponse::getErrorResponse("File already exists in current directory. Please rename file or put under different directory");
+                }
+                FileUtils::writeFile($new_path, "");
+                rename($course_material->getPath(), $new_path);
+                $course_material->setPath($new_path);
+            }
+            elseif ($course_material->isLink()) {
+                $course_material->setUrl($_POST['file_path']);
+            }
+        }
+
+        if (isset($_POST['display_name'])) {
+            $display_name = $_POST['display_name'];
+            $path = $course_material->getPath();
+            $dirs = explode("/", $path);
+            array_pop($dirs);
+            $path = implode("/", $dirs);
+            if ($course_material->isLink() && $display_name !== $course_material->getDisplayName()) {
```

Comment:
Is the display_name changing the location of where the file is getting stored? Is it not possible to just always keep the path the same and just visually change the text to be whatever the display name is? Or is the real filepath supposed to be hidden from users?

Reply:
The display name was to solve the problem where professors wanted to name their files differently than what is being shown to students. So they would see file1-v2 and students would see file1. Display_name used to be called url_title as we didn't want to show the whole link, but a description of the link. Now it serves two purposes. For files, it is the name students/ta/mentors see. For links, it is the link title. I changed the path when a url is being edited here as that was the original behavior. 

This is what editing a link previously looked liked. I changed it so editing a link and file would have same UI. 
![urllink](https://user-images.githubusercontent.com/29582421/164278270-585f42d3-8a25-487e-9a1b-84a7b2e3e739.PNG)

Reply:
Right but why would the real location of a URL or file need to be changed if only the display name is getting updated?

Reply:
This is only the case for links. I'm not sure what the original design decisions behind this was. For links, the path column would be `/var/local/submitty/courses/s22/sample/uploads/course_materials/link-display_name`, the url column would have the link, and the display_name (originally url_title) column would store the display_name.
![table](https://user-images.githubusercontent.com/29582421/164307311-c0c0cfef-5226-4c33-9c6d-93b0484c4bcd.PNG)